File: ascii_fields/animations/lava.py
import math

from ..core import Animation, clamp, render_field
from ..noise import fbm


class LavaAnimation(Animation):
  """A molten lava surface with convection cells, hot bubbles, dark cooled
  cracks running across the crust, and occasional bursting rings."""

  default_theme = "lava"

  thresholds = [
    (0.07, " "), (0.16, "."), (0.27, ":"), (0.39, "-"), (0.52, "="),
    (0.65, "+"), (0.78, "*"), (0.90, "#"), (1.01, "@"),
  ]

  BUBBLES = (
    (0.18, 0.82, 5.0, 0.00, 0.08),
    (0.34, 0.76, 6.4, 0.23, 0.06),
    (0.56, 0.84, 4.6, 0.47, 0.09),
    (0.74, 0.78, 7.2, 0.68, 0.055),
    (0.88, 0.86, 5.7, 0.81, 0.07),
  )

  def render(self, width, height, elapsed, phase, options):
    contrast = options.contrast
    # Precompute per-bubble state once per frame (was being recomputed per cell).
    bubble_state = []
    for bx, by, period, offset, radius in self.BUBBLES:
      local = (elapsed / period + offset) % 1.0
      rise = by - local * 0.42
      wobble = 0.035 * math.sin(elapsed * 1.3 + offset * 9.0)
      bx_eff = bx + wobble
      bursting = local > 0.78
      burst = (local - 0.78) / 0.22 if bursting else 0.0
      ring_r = 0.45 + burst * 2.1 if bursting else 0.0
      ring_decay = (1.0 - burst) ** 1.5 if bursting else 0.0
      spark_phase = offset * 140.0
      bubble_state.append((bx_eff, rise, radius, local, bursting, ring_r, ring_decay, spark_phase))

    grid = []
    for row in range(height):
      v = row / max(1, height - 1)
      one_minus_v = 1.0 - v
      sin_v7 = math.sin(v * 7.0) * 1.6           # row-only term in cracks
      drift = elapsed * 0.18
      line = []
      for col in range(width):
        u = col / max(1, width - 1)
        convection = fbm(u * 3.0 + elapsed * 0.10, v * 2.4 - drift, octaves=4)
        cracks = abs(math.sin(u * 18.0 + sin_v7 - elapsed * 0.8))
        # base heat with wider contrast and dark cooled crust along cracks
        heat = 0.16 + 0.52 * convection + 0.22 * one_minus_v
        heat -= max(0.0, cracks - 0.55) * 0.40   # cracks are *dark*

        for bx_eff, rise, radius, local, bursting, ring_r, ring_decay, spark_phase in bubble_state:
          inv_r = 1.0 / max(0.001, radius)
          dx = (u - bx_eff) * inv_r
          dy = (v - rise) * inv_r / 0.55
          dist2 = dx * dx + dy * dy
          bubble = math.exp(-dist2 * 1.7)
          heat += bubble * (0.34 + 0.22 * local)
          if bursting:
            dist = math.sqrt(dist2)
            ring = math.exp(-((dist - ring_r) ** 2) / 0.08) * ring_decay
            sparks = max(0.0, math.sin((u + spark_phase * 0.01) * 140.0 + elapsed * 9.0))
            heat += ring * (0.60 + 0.30 * sparks)

        line.append(clamp(heat * contrast))
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/lava.rs
use rayon::prelude::*;

use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};
use crate::noise::fbm;

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "lava" };
const TH: &[(f64, char)] = &[
  (0.07, ' '),
  (0.16, '.'),
  (0.27, ':'),
  (0.39, '-'),
  (0.52, '='),
  (0.65, '+'),
  (0.78, '*'),
  (0.90, '#'),
  (1.01, '@'),
];
// (bx, by, period, offset, radius)
const BUBBLES: &[(f64, f64, f64, f64, f64)] = &[
  (0.18, 0.82, 5.0, 0.00, 0.08),
  (0.34, 0.76, 6.4, 0.23, 0.06),
  (0.56, 0.84, 4.6, 0.47, 0.09),
  (0.74, 0.78, 7.2, 0.68, 0.055),
  (0.88, 0.86, 5.7, 0.81, 0.07),
];

struct BubbleState {
  bx_eff: f64,
  rise: f64,
  inv_rx: f64, // 1 / radius, hoisted out of the per-cell loop
  inv_ry: f64, // 1 / (radius * 0.55), same
  local: f64,
  bursting: bool,
  ring_r: f64,
  ring_decay: f64,
  spark_x_scale: f64, // 140.0 (per-cell coefficient on u)
  spark_phase: f64,   // 140 * 0.01 * offset, pre-shifted
}

pub struct Lava;

impl Animation for Lava {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width;
    let h = ctx.height;
    let contrast = ctx.options.contrast;
    let bubble_state: Vec<BubbleState> = BUBBLES
      .iter()
      .map(|&(bx, by, period, offset, radius)| {
        let local = ((ctx.elapsed / period + offset) % 1.0 + 1.0) % 1.0;
        let rise = by - local * 0.42;
        let wobble = 0.035 * (ctx.elapsed * 1.3 + offset * 9.0).sin();
        let bx_eff = bx + wobble;
        let bursting = local > 0.78;
        let burst = if bursting { (local - 0.78) / 0.22 } else { 0.0 };
        let ring_r = if bursting { 0.45 + burst * 2.1 } else { 0.0 };
        let ring_decay = if bursting { (1.0 - burst).powf(1.5) } else { 0.0 };
        let inv_rx = 1.0 / radius.max(0.001);
        let inv_ry = inv_rx / 0.55;
        BubbleState {
          bx_eff,
          rise,
          inv_rx,
          inv_ry,
          local,
          bursting,
          ring_r,
          ring_decay,
          spark_x_scale: 140.0,
          spark_phase: offset * 140.0 * 0.01,
        }
      })
      .collect();

    // Frame-invariant terms hoisted out of the row/col loops.
    let drift = ctx.elapsed * 0.18;
    let elapsed_conv = ctx.elapsed * 0.10;
    let elapsed_crack = ctx.elapsed * 0.8;
    let elapsed_spark = ctx.elapsed * 9.0;
    let mut grid = vec![0.0_f64; w * h];
    let dw = (w.saturating_sub(1)).max(1) as f64;
    let dh = (h.saturating_sub(1)).max(1) as f64;
    let bubble_state = &bubble_state; // shared, captured by ref into closures
    grid.par_chunks_mut(w).enumerate().for_each(|(row, row_slice)| {
      let v = row as f64 / dh;
      let one_minus_v = 1.0 - v;
      let sin_v7 = (v * 7.0).sin() * 1.6;
      let v_drift = v * 2.4 - drift; // fbm Y argument is row-only
      let base_heat = 0.16 + 0.22 * one_minus_v; // row-only part of `heat`
      for col in 0..w {
        let u = col as f64 / dw;
        let convection = fbm(u * 3.0 + elapsed_conv, v_drift, 4);
        let cracks = (u * 18.0 + sin_v7 - elapsed_crack).sin().abs();
        let mut heat = base_heat + 0.52 * convection;
        heat -= (cracks - 0.55).max(0.0) * 0.40;
        for b in bubble_state {
          let dx = (u - b.bx_eff) * b.inv_rx;
          let dy = (v - b.rise) * b.inv_ry;
          let dist2 = dx * dx + dy * dy;
          let bubble = (-dist2 * 1.7).exp();
          heat += bubble * (0.34 + 0.22 * b.local);
          if b.bursting {
            let dist = dist2.sqrt();
            let ring = (-((dist - b.ring_r).powi(2)) / 0.08).exp() * b.ring_decay;
            let sparks = ((u + b.spark_phase) * b.spark_x_scale + elapsed_spark).sin().max(0.0);
            heat += ring * (0.60 + 0.30 * sparks);
          }
        }
        row_slice[col] = clamp(heat * contrast);
      }
    });
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
